keps/sig-api-machinery/20180415-crds-to-ga.md

```diff
@@ -0,0 +1,204 @@
+
+### Generator exists for CRD Validation Schema v3 (Kubebuilder)
+
+See [kubebuilder’s
+crd-gen](https://github.com/kubernetes-sigs/controller-tools/tree/master/cmd/crd)
+and a more general
+[crd-schema-gen](https://github.com/kubernetes-sigs/controller-tools/pull/183),
+to be integrated into kubebuidler’s controller-tools.
+
+### CustomResourceWebhookConversion API is GA ready
+
+Currently conversion webhooks are v1alpha1. Our goal is to move to v1beta1 in
+the Kubernetes 1.15 release, and then graduate this to GA as part of the CRD to
+GA graduation; we won't track this as a separate feature from CRDs once it's in
+v1beta1.
+
+Conversion webhooks to beta tasks:
+- ConversionReview API v1 (incl. fields that need to be updated for v1beta1 or v1)
+- Add conversion webhook tests enumerated in below [Test Plan](#test-plan)
+* Add comprehensive metrics to conversion webhooks
+
+### CustomResourceSubresources API is GA ready
+
+Currently custom resource subresources are v1beta1 and provide support for the
+/status and /scale subresources.  We plan to graduate this to GA as part of the
+CRD to GA graduation. See [Arbitrary subresources as Post-GA
+task](#arbitrary-subresources-as-post-ga-task) details on post GA tasks.
+
+### v1 API
+
+The CRD `v1` API will be the same as the `v1beta1` but with all changes to the API from the GA tasks:
+
+* Rename misnamed json field [JSONPath](https://github.com/kubernetes/kubernetes/blob/06bc7e3e0026ea25065f59f4bd305c0b7dbbc145/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go#L226-L227) to `jsonPath`
+* [Replace top-level fields with per version fields](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/customresource-conversion-webhook.md#top-level-fields-to-per-version-fields)
+* Restrict OpenAPI per [Vanilla OpenAPI Subset Design](https://docs.google.com/document/d/1pcGlbmw-2Y0JJs9hsYnSBXamgG9TfWtHY6eh80zSTd8)
+
+## Test Plan
+
+### Integration tests for GA
+
+TODO: complete this list
+
+* [Pruning for CustomResources KEP](https://github.com/kubernetes/enhancements/pull/709)
+  * TODO
+* [Vanilla OpenAPI Subset Design](https://docs.google.com/document/d/1pcGlbmw-2Y0JJs9hsYnSBXamgG9TfWtHY6eh80zSTd8)
+  * TODO
+* CRD versioning/conversion (https://github.com/kubernetes/kubernetes/issues/64136):
+  * Ensure what is persisted in etcd matches the storage version
+  * Set up a CRD, persist some data, changed the version, and ensure the previously persisted data is readable
+  * Ensure discovery docs track a CRD through creation, version addition, version removal, and deletion
+  * Ensure `spec.served` is respected
+
+### e2e tests for GA
+
+TODO: complete this list
+
+* Ensure v1beta1 and v1 CRDs are round trippable
+* Ensure CR watch observes events on create/update/delete
+* Ensure a CR watch is correctly terminated if CRD is deleted or updated
+
+### Conformance plan for GA
+
+TODO: complete this list
+
+* Promote existing e2e tests:
```

Comment:
Is this planned to happen simultaneously with v1 GA?

Reply:
Probably should be a follow up given the conformance process. I'll add a line item to the post GA list.